### ISSUE TITLE
Least squares setup for model merge

### ISSUE THREAD

**Issue by johnmullan** [adobe-research/custom-diffusion#10]
Hi, when iterating through the attention layers to form Vtarget1, how come we are able to reuse the same text embedding (when multiplying the weights [here](https://github.com/adobe-research/custom-diffusion/blob/main/src/composenW.py#L165)) throughout all the layers? Wouldn't this embedding get transformed as it propagates through the network? 

How hard would this be to extend this concept to not only the attention layers? I'm wanting to merge two fine tuned models for clip and unet, trained on all layers.

Also I think[ line 91 in composenW.py](https://github.com/adobe-research/custom-diffusion/blob/main/src/composenW.py#L91) is a bug where prompts param isn't used?

**Comment by nupurkmr9** [comment]
Hi, 

Thanks for pointing out the mistake in line 91. I have corrected it now. 
Regarding reusing the same text embedding [here](https://github.com/adobe-research/custom-diffusion/blob/main/src/composenW.py#L165), this is because the same text embedding is input to all the key, value projection matrices in the attention layers. 

How to merge two fully fine-tuned networks is an interesting question. I doubt linear least square constraints will be sufficient for that as -- (1) obtaining input output constraint is more difficult for all weights because of its dependency on the sampled noise and previous layers, and (2) non-linearities in the network. 
Maybe it can still be solved through gradient descent optimization on the constraints. One of the relevant works might be [this](https://arxiv.org/abs/2007.15646). 
Maybe fine-tuning can be done such that simple addition of weights leads to model merging. I don't have any specific insight but its an exciting direction to explore :)

**Comment by johnmullan** [comment]
Thanks, that makes perfect sense regarding the text embedding input.

Thanks for the link to the paper. This seams a much more logical way to go - especially due to the non-linearity in the network.
